roles/roles-utils/network-helpers/src/noise_stream.rs

```diff
@@ -0,0 +1,186 @@
+                            let transport_state = state.step_2(payload)?;
+                            state = transport_state;
+                            break;
+                        }
+                        Err(Error::CodecError(codec_sv2::Error::MissingBytes(_))) => {
+                            debug!("Waiting for more bytes during handshake");
+                        }
+                        Err(e) => {
+                            error!("Handshake failed with upstream: {:?}", e);
+                            return Err(e);
+                        }
+                    }
+                }
+            }
+            HandshakeRole::Responder(_) => {
+                let mut initiator_state = codec_sv2::State::not_initialized(&role);
+
+                loop {
+                    match receive_message(&mut reader, &mut initiator_state, &mut decoder).await {
+                        Ok(first_msg) => {
+                            debug!("First handshake message received");
+                            let handshake_frame: HandShakeFrame = first_msg
+                                .try_into()
+                                .map_err(|_| Error::HandshakeRemoteInvalidMessage)?;
+                            let payload: [u8; ELLSWIFT_ENCODING_SIZE] = handshake_frame
+                                .get_payload_when_handshaking()
+                                .try_into()
+                                .map_err(|_| Error::HandshakeRemoteInvalidMessage)?;
+                            let (second_msg, transport_state) = state.step_1(payload)?;
+                            send_message(&mut writer, second_msg.into(), &mut state, &mut encoder)
+                                .await?;
+                            debug!("Second handshake message sent");
+                            state = transport_state;
+                            break;
+                        }
+                        Err(Error::CodecError(codec_sv2::Error::MissingBytes(_))) => {
+                            debug!("Waiting for more bytes during handshake");
+                        }
+                        Err(e) => {
+                            error!("Handshake failed with downstream: {:?}", e);
+                            return Err(e);
+                        }
+                    }
+                }
+            }
+        };
+        Ok(Self {
+            reader: NoiseTcpReadHalf {
+                reader,
+                decoder,
+                state: state.clone(),
+            },
+            writer: NoiseTcpWriteHalf {
+                writer,
+                encoder,
+                state,
+            },
+        })
+    }
+}
+
+impl<Message> NoiseTcpWriteHalf<Message>
+where
+    Message: Serialize + for<'a> Deserialize<'a> + GetSize + Send + 'static,
+{
+    pub async fn write_frame(&mut self, frame: StandardEitherFrame<Message>) -> Result<(), Error> {
```

Comment:
We should add a doc comment here that this is not cancellation safe as tokio does for `.write_all`